src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs

```diff
@@ -113,21 +113,24 @@ public class Frame<TContext> : Frame
                                 await FireOnCompleted();
                             }
 
-                            _application.DisposeContext(context, _applicationException);
-                        }
-
-                        // If _requestAbort is set, the connection has already been closed.
-                        if (Volatile.Read(ref _requestAborted) == 0)
-                        {
-                            ResumeStreams();
-
-                            if (_keepAlive)
+                            // If _requestAbort is set, the connection has already been closed.
+                            if (Volatile.Read(ref _requestAborted) == 0)
                             {
-                                // Finish reading the request body in case the app did not.
-                                await messageBody.Consume();
+                                ResumeStreams();
+
+                                if (_keepAlive)
+                                {
+                                    // Finish reading the request body in case the app did not.
+                                    await messageBody.Consume();
+                                }
+
+                                // ProduceEnd() must be called before _application.DisposeContext(), to ensure
+                                // HttpContext.Response.StatusCode is correct set when
+                                // IHttpContextFactory.Dispose(HttpContext) is called.
+                                await ProduceEnd();
                             }
 
-                            await ProduceEnd();
+                            _application.DisposeContext(context, _applicationException);
```

Comment:
Currently, if the app calls `HttpContext.Abort()`, then `HttpContext.Response.StatusCode` will be `200` when it's disposed (which appears in the logs).  I agree this seems wrong.  However, if the app aborts the request then no status code is sent to the client, so the status of the request is technically undefined.

What should `HttpContext.Response.StatusCode` be in this case?  We can set it to `500` which is arguably better than `200`, but something like `null` or `-1` (property is an `int` rather than `HttpStatusCode`) might be more appropriate.